usersync/cookie.go

```diff
@@ -20,70 +19,148 @@ const uidTTL = 14 * 24 * time.Hour
 
 // Cookie is the cookie used in Prebid Server.
 //
-// To get an instance of this from a request, use ParseCookieFromRequest.
-// To write an instance onto a response, use SetCookieOnResponse.
+// To get an instance of this from a request, use ReadCookie.
+// To write an instance onto a response, use WriteCookie.
 type Cookie struct {
-	uids   map[string]uidWithExpiry
+	uids   map[string]UIDEntry
 	optOut bool
 }
 
-// uidWithExpiry bundles the UID with an Expiration date.
-// After the expiration, the UID is no longer valid.
-type uidWithExpiry struct {
+// UIDEntry bundles the UID with an Expiration date.
+type UIDEntry struct {
 	// UID is the ID given to a user by a particular bidder
 	UID string `json:"uid"`
 	// Expires is the time at which this UID should no longer apply.
 	Expires time.Time `json:"expires"`
 }
 
-// ParseCookieFromRequest parses the UserSyncMap from an HTTP Request.
-func ParseCookieFromRequest(r *http.Request, cookie *config.HostCookie) *Cookie {
-	if cookie.OptOutCookie.Name != "" {
-		optOutCookie, err1 := r.Cookie(cookie.OptOutCookie.Name)
-		if err1 == nil && optOutCookie.Value == cookie.OptOutCookie.Value {
-			pc := NewCookie()
-			pc.SetOptOut(true)
-			return pc
-		}
+// NewCookie returns a new empty cookie.
+func NewCookie() *Cookie {
+	return &Cookie{
+		uids: make(map[string]UIDEntry),
 	}
-	var parsed *Cookie
-	uidCookie, err2 := r.Cookie(uidCookieName)
-	if err2 == nil {
-		parsed = ParseCookie(uidCookie)
-	} else {
-		parsed = NewCookie()
+}
+
+// ReadCookie reads the cookie from the request
+func ReadCookie(r *http.Request, decoder Decoder, host *config.HostCookie) *Cookie {
+	if hostOptOutCookie := checkHostCookieOptOut(r, host); hostOptOutCookie != nil {
+		return hostOptOutCookie
+	}
+
+	// Read cookie from request
+	cookieFromRequest, err := r.Cookie(uidCookieName)
+	if err != nil {
+		return NewCookie()
 	}
-	// Fixes #582
-	if uid, _, _ := parsed.GetUID(cookie.Family); uid == "" && cookie.CookieName != "" {
-		if hostCookie, err := r.Cookie(cookie.CookieName); err == nil {
-			parsed.TrySync(cookie.Family, hostCookie.Value)
+	decodedCookie := decoder.Decode(cookieFromRequest.Value)
+
+	return decodedCookie
+}
+
+// PrepareCookieForWrite ejects UIDs as long as the cookie is too full
+func (cookie *Cookie) PrepareCookieForWrite(cfg *config.HostCookie, encoder Encoder) (string, error) {
+	uuidKeys := sortUIDs(cookie.uids)
+
+	i := 0
+	isCookieTooBig := true
+
+	for isCookieTooBig && len(cookie.uids) > 0 {
```

Comment:
If `isCookieTooBig` is negative, we would have exited the `for` statement in the previous loop at line 74. We probably don't need this variable as a condition in the `for` statement at all:
```
 60   // PrepareCookieForWrite ejects UIDs as long as the cookie is too full
 61   func (cookie *Cookie) PrepareCookieForWrite(cfg *config.HostCookie, encoder Encoder) (string, error) {
 62       uuidKeys := sortUIDs(cookie.uids)
 63  
 64       i := 0
 65 -     isCookieTooBig := true
 66  
 67 -     for isCookieTooBig && len(cookie.uids) > 0 {
    +     for len(cookie.uids) > 0 {
 68           encodedCookie, err := encoder.Encode(cookie)
 69           if err != nil {
 70               return encodedCookie, nil
 71           }
 72  
 73 -         isCookieTooBig = len(encodedCookie) > cfg.MaxCookieSizeBytes && cfg.MaxCookieSizeBytes > 0
    +         isCookieTooBig := len(encodedCookie) > cfg.MaxCookieSizeBytes && cfg.MaxCookieSizeBytes > 0
 74           if !isCookieTooBig {
 75               return encodedCookie, nil
 76           }
 77  
 78           uidToDelete := uuidKeys[i]
 79           delete(cookie.uids, uidToDelete)
 80  
 81           i++
 82       }
 83       return "", nil
 84   }
usersync/cookie.go
```